File: typegraph/typegraph/runtimes/graphql.py
# Copyright Metatype OÜ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.

from typing import Optional, Tuple

from attrs import frozen

from typegraph import effects
from typegraph import types as t
from typegraph.effects import Effect
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.utils.attrs import always, required


@frozen
class GraphQLRuntime(Runtime):
    endpoint: str
    runtime_name: str = always("graphql")

    def query(
        self, inp: t.struct, out: t.typedef, *, path: Optional[Tuple[str, ...]] = None
    ):
        if path is not None and len(path) == 0:
            raise Exception("Unexpected empty path")
        return t.func(inp, out, QueryMat(self, path))

    def mutation(
        self,
        inp: t.struct,
        out: t.typedef,
        path: Optional[Tuple[str, ...]] = None,
        **kwargs
    ):
        if path is not None and len(path) == 0:
            raise Exception("Unexpected empty path")
        return t.func(inp, out, MutationMat(self, path, **kwargs))


@frozen
class QueryMat(Materializer):
    runtime: Runtime
    path: Optional[Tuple[str]]
    materializer_name: str = always("query")
    effect: Effect = always(effects.none())


@frozen
class MutationMat(Materializer):
    runtime: Runtime
    path: Optional[Tuple[str]]
    materializer_name: str = always("mutation")
    effect: Effect = required()

File: typegraph/deno/src/runtimes/graphql.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import { Effect } from "../../gen/exports/metatype-typegraph-runtimes.d.ts";
import * as t from "../types.ts";
import { runtimes } from "../wit.ts";
import { Materializer, Runtime } from "./mod.ts";
import * as effects from "../effects.ts";

export class GraphQLRuntime extends Runtime {
  constructor(private endpoint: string) {
    super(runtimes.registerGraphqlRuntime({
      endpoint,
    }));
  }

  query<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(inp: I, out: O, path: string[] = []): t.Func<P, I, O, QueryMat> {
    const matId = runtimes.graphqlQuery({
      runtime: this._id,
      effect: { tag: "none" },
    }, {
      path,
    });
    const mat: QueryMat = {
      _id: matId,
      path,
    };
    return t.func(inp, out, mat);
  }

  mutation<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(
    inp: I,
    out: O,
    effect: Effect,
    path: string[] = [],
  ): t.Func<P, I, O, MutationMat> {
    const matId = runtimes.graphqlMutation({
      runtime: this._id,
      effect,
    }, {
      path,
    });
    const mat: MutationMat = {
      _id: matId,
      path,
    };
    return t.func(inp, out, mat);
  }
}

interface QueryMat extends Materializer {
  path: string[];
}

interface MutationMat extends QueryMat {
}
